lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Random.hs

```diff
@@ -369,3 +449,46 @@ decodeNestedBytes dec bytes =
         Right ("", res) -> pure res
         Right _ -> fail "Leftovers when decoding nested bytes"
         _ -> fail "Could not decode nested bytes"
+
+{-------------------------------------------------------------------------------
+                                   Passphrase
+-------------------------------------------------------------------------------}
+
+-- | Re-encrypt the private key using a different passphrase, and regenerate
+-- subkeys.
+--
+-- **Important**:
+-- This function doesn't check that the old passphrase is correct! Caller is
+-- expected to have already checked that. Using an incorrect passphrase here
+-- will lead to very bad thing.
+changePassphraseRnd
+    :: Passphrase "encryption-old"
+    -> Passphrase "encryption-new"
+    -> RndKey purpose XPrv
+    -> RndKey purpose XPrv
+changePassphraseRnd (Passphrase oldPwd) (Passphrase newPwd) k = k'
+  where
+    k' = deriveRndKey (Passphrase newPwd) $ Key xprv'
+    xprv' = xPrvChangePass oldPwd newPwd (getKey $ masterKey k)
+
+{-------------------------------------------------------------------------------
+                              WalletKey functions
+-------------------------------------------------------------------------------}
+
+-- | Extract the public key part of a private key.
+publicKeyRnd
+    :: RndKey level XPrv
+    -> RndKey level XPub
+publicKeyRnd k = k
+    { masterKey = toXPub <$> masterKey k
```

Comment:
the same remark about `toXPub` as above

Reply:
ok, I have only now detected that `publicKey` does not take form like this anymore